builder/docker/docker.go

```diff
@@ -0,0 +1,350 @@
+		return true, os.Remove(DockSock)
+	}
+	return false, nil
+}
+
+// CreateClient creates a new Docker client.
+//
+// Params:
+// 	- url (string): The URI to the Docker daemon. This defaults to the UNIX
+// 		socket /var/run/docker.sock.
+//
+// Returns:
+// 	- *docker.Client
+//
+func CreateClient(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	path := p.Get("url", "unix:///var/run/docker.sock").(string)
+
+	return docli.NewClient(path)
+}
+
+// Start starts a Docker daemon.
+//
+// This assumes the presence of the docker client on the host. It does not use
+// the API.
+func Start(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	dargs := []string{
+		"-d",
+		"--bip=172.19.42.1/16",
+		"--insecure-registry",
+		"10.0.0.0/8",
+		"--insecure-registry",
+		"172.16.0.0/12",
+		"--insecure-registry",
+		"192.168.0.0/16",
+		"--insecure-registry",
+		"100.64.0.0/10",
+		//"--storage-driver=overlay", // Add this dynamically.
+	}
+
+	// There is probably a better way to do this. I'm not sure whether the
+	// original intent of the mkdir was to accomplish something, or just check
+	// whether overlay is supported.
+	if err := os.MkdirAll("/", 0700); err == nil {
+
+		cmd := exec.Command("findmnt", "--noheadings", "--output", "FSTYPE", "--target", "/")
+
+		if out, err := cmd.Output(); err == nil && strings.TrimSpace(string(out)) == "overlay" {
+			dargs = append(dargs, "--storage-driver=overlay")
+		} else {
+			log.Infof(c, "File system type: '%s' (%v)", out, err)
+		}
+	}
+
+	//# spawn a docker daemon to run builds
+	//docker -d --bip=172.19.42.1/16 $DRIVER_OVERRIDE --insecure-registry 10.0.0.0/8 --insecure-registry 172.16.0.0/12
+	// --insecure-registry 192.168.0.0/16 --insecure-registry 100.64.0.0/10 &
+	//DOCKER_PID=$!
+
+	log.Infof(c, "Starting docker with %s", strings.Join(dargs, " "))
+	cmd := exec.Command("docker", dargs...)
+	if err := cmd.Start(); err != nil {
+		log.Errf(c, "Failed to start Docker. %s", err)
+		return -1, err
+	}
+	// Get the PID and return it.
+	return cmd.Process.Pid, nil
+}
+
+// WaitForStart delays until Docker appears to be up and running.
+//
+// Params:
+// 	- client (*docker.Client): Docker client.
+// 	- timeout (time.Duration): Time after which to give up.
+//
+// Returns:
+// 	- boolean true if the server is up.
+func WaitForStart(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	ok, missing := p.RequiresValue("client")
+	if !ok {
+		return nil, &cookoo.FatalError{"Missing required fields: " + strings.Join(missing, ", ")}
+	}
+	cli := p.Get("client", nil).(*docli.Client)
+	timeout := p.Get("timeout", 30*time.Second).(time.Duration)
+
+	keepon := true
+	timer := time.AfterFunc(timeout, func() {
+		keepon = false
+	})
+
+	for keepon == true {
+		if err := cli.Ping(); err == nil {
+			timer.Stop()
+			log.Infof(c, "Docker is running.")
+			return true, nil
+		}
+		time.Sleep(time.Second)
+	}
```

Comment:
line 121-133. I feel that this can be done much cleaner having two time channels. A ticker of 1 second (or less) to ping docker, if docker is running stop all timers and return. The other timer is a simple timeout timer. `time.AfterFunc` is cool but I think it makes this code quite unreadable.